src/traces/splom/index.js

```diff
@@ -36,19 +36,39 @@ function calc(gd, trace) {
     // only differ here for log axes, pass ldata to createMatrix as 'data'
     var cdata = opts.cdata = [];
     var ldata = opts.data = [];
-    var i, k, dim;
+    var i, k, dim, xa, ya;
+
+    function makeCalcdata(ax, dim) {
+        // call makeCalcdata with fake input
+        var ccol = ax.makeCalcdata({
+            v: dim.values,
+            vcalendar: trace.calendar
+        }, 'v');
+
+        for(var i = 0; i < ccol.length; i++) {
+            ccol[i] = ccol[i] === BADNUM ? NaN : ccol[i];
+        }
+        cdata.push(ccol);
+        ldata.push(ax.type === 'log' ? Lib.simpleMap(ccol, ax.c2l) : ccol);
+    }
 
     for(i = 0; i < dimensions.length; i++) {
         dim = dimensions[i];
 
         if(dim.visible) {
-            var axId = trace._diag[i][0] || trace._diag[i][1];
-            var ax = AxisIDs.getFromId(gd, axId);
-            if(ax) {
-                var ccol = makeCalcdata(ax, trace, dim);
-                var lcol = ax.type === 'log' ? Lib.simpleMap(ccol, ax.c2l) : ccol;
-                cdata.push(ccol);
-                ldata.push(lcol);
+            xa = AxisIDs.getFromId(gd, trace._diag[i][0]);
+            ya = AxisIDs.getFromId(gd, trace._diag[i][1]);
+
+            if(xa) {
+                makeCalcdata(xa, dim);
+                if(ya && ya.type === 'category') {
+                    ya._categories = xa._categories.slice();
+                }
+            } else if(ya) {
+                makeCalcdata(ya, dim);
+                if(xa && xa.type === 'category') {
+                    xa._categories = ya._categories.slice();
+                }
```

Comment:
How could you get here? if `xa` exists you'll skip the `else if(ya)` block entirely.

But this raises an interesting (if possibly nonsensical) point - what if `xa.type !== ya.type`? I guess  this could happen if one is numeric or date and the other is categorical, or if one is linear and the other is log. If the answer is "all hell breaks loose", which wouldn't surprise me, it may be appropriate to just log a warning and hide that dimension entirely. That seems like an abuse of the concept of a splom.

Reply:
> a users sets ax.type in the layout
> ... [doesn't] happen by accident.

Hmm could happen in the editor though. @nicolaskruchten @VeraZab not sure how best to handle this. If we add [`dimension.type`](https://github.com/plotly/plotly.js/pull/2899#discussion_r209731332) we should steer people there somehow when there's a splom on that axis. But if they still set an axis type directly would we want to find the corresponding opposite axis in the splom and set it to match, to avoid this error?

Reply:
hoo boy. Well right now we don't support SPLOMs yet, and I'm thinking that when we do we should lock down axes/subplots quite tightly so that people can't break them too much :)